### PR TITLE
Have &cxxname automatically prefix IDs with `::`

### LINKED ISSUE #1285: Have `&cxxname` automatically prefix IDs with `::`.
This is to avoid scope lookup errors inside namespace that happen to have a different definition for the same ID. Then also remove the `::` from `cxxname` in our library files.

Update TODO in `doc/extending.rst` once done.

### PR DESCRIPTION
**Pull request by bbannier**
Closes #1285.

### CODE REVIEW

File: hilti/runtime/include/logging.h [hunk 1]
```diff
@@ -99,5 +89,17 @@ inline void setLocation(const char* l = nullptr) {
         context->source_location = l;
 }
 
+/**
+ * Prints a string, or a runtime value, to a specific debug stream. This is a
+ * macro wrapper around `debug::detail::print(*)` that avoids evaluation of
```
Comment (rsmmr) [inline hilti/runtime/include/logging.h:94]:
```suggestion
 * wrapper around `debug::detail::print(*)` that avoids evaluation of
```

File: hilti/toolchain/src/compiler/visitors/normalizer.cc [hunk 2]
```diff
@@ -142,6 +143,19 @@ struct VisitorNormalizer : public visitor::PreOrder<void, VisitorNormalizer> {
...
             type::Struct::setSelf(&p.node);
     }
+
+    void operator()(const Attribute& n, position_t p) {
+        // Normalize values passed as `&cxxname` so they always are interpreted as FQNs by enforcing leading `::`.
+        if ( const auto& tag = n.tag(); tag == "&cxxname" && n.hasValue() ) {
+            if ( const auto& value = n.valueAsString(); value && ! util::startsWith(*value, "::") ) {
+                auto new_n = Attribute(tag, builder::string(util::fmt("::%s", *value)));
+                HILTI_DEBUG(logging::debug::Normalizer,
+                            util::fmt("[%s] %s -> %s (%s)", *value, n, new_n, n.value().location()));
```
Comment (rsmmr) [inline hilti/toolchain/src/compiler/visitors/normalizer.cc:153]: Suggest to add new `log_change` method here for attributes.

File: hilti/toolchain/src/ast/types/library.cc [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+#include <utility>
+
+#include <hilti/base/util.h>
+
+using namespace hilti;
+using namespace hilti::type;
+
+static std::string normalize(std::string cxx_name) {
+    if ( cxx_name.find(' ') == std::string::npos && ! util::startsWith(cxx_name, "::") )
```
Comment (rsmmr) [inline hilti/toolchain/src/ast/types/library.cc:13]: What's the `find(' ')` for?

**Review by rsmmr — DISMISSED** [review]
Nice, this has been bothering me for a while!

File: hilti/toolchain/src/ast/types/library.cc (hunk 3 above)
Comment (bbannier) [reply hilti/toolchain/src/ast/types/library.cc:13]: This is a dedicated workaround for `__library_type("const hilti::rt::TypeInfo*")`. We can get rid of treatment of spaces if we use `__library_type("hilti::rt::TypeInfo* const")` instead; this seems fine to me as `__library_type` is an internal API anyway.

Comment (bbannier) [reply hilti/toolchain/src/ast/types/library.cc:13]: Did just that.

**Pull request merged**